File: js/examples/keypair_example.ts
/**
 * TOPAY-Z512 Key Pair Example
 * 
 * This example demonstrates how to generate and use key pairs in TOPAY-Z512.
 * It shows key generation, public key derivation, and hex conversion.
 */

import { KeyPair, PrivateKey, PublicKey, generateKeyPair } from '../src/keypair';

console.log('TOPAY-Z512 Key Pair Example\n');

// Generate a new key pair
const keypair = generateKeyPair();

console.log('Generated Key Pair:');
console.log(`Private Key: ${keypair.privateKey.toHex()}`);
console.log(`Public Key:  ${keypair.publicKey.toHex()}\n`);

// Demonstrate deriving a public key from a private key
const derivedPublicKey = PublicKey.fromPrivateKey(keypair.privateKey);
console.log(`Derived Public Key: ${derivedPublicKey.toHex()}`);
console.log(`Keys match: ${derivedPublicKey.toHex() === keypair.publicKey.toHex()}\n`);

// Demonstrate hex conversion
const privateHex = keypair.privateKey.toHex();
const recoveredPrivateKey = PrivateKey.fromHex(privateHex);

const publicHex = keypair.publicKey.toHex();
const recoveredPublicKey = PublicKey.fromHex(publicHex);

console.log('Hex Conversion Test:');
console.log(`Private Key Recovered: ${privateHex === recoveredPrivateKey.toHex()}`);
console.log(`Public Key Recovered:  ${publicHex === recoveredPublicKey.toHex()}`);

// Create a new key pair from existing keys
const newKeypair = new KeyPair(recoveredPrivateKey, recoveredPublicKey);
console.log('\nRecreated Key Pair:');
console.log(`Private Key: ${newKeypair.privateKey.toHex()}`);
console.log(`Public Key:  ${newKeypair.publicKey.toHex()}`);

File: go/examples/keypair_example.go
// TOPAY-Z512 Key Pair Example
//
// This example demonstrates how to generate and use key pairs in TOPAY-Z512.
// It shows key generation, public key derivation, and hex conversion.
package examples

import (
	"fmt"
	"log"

	"github.com/topay-foundation/topayz512/pkg/keypair"
)

// RunKeypair demonstrates the key pair functionality
func RunKeypair() {
	fmt.Println("TOPAY-Z512 Key Pair Example")

	// Generate a new key pair
	kp, err := keypair.GenerateKeyPair()
	if err != nil {
		log.Fatalf("Failed to generate key pair: %v", err)
	}

	fmt.Println("Generated Key Pair:")
	fmt.Printf("Private Key: %s\n", kp.PrivateKey.String())
	fmt.Printf("Public Key:  %s\n\n", kp.PublicKey.String())

	// Demonstrate deriving a public key from a private key
	derivedPublicKey := keypair.DerivePublicKey(kp.PrivateKey)
	fmt.Printf("Derived Public Key: %s\n", derivedPublicKey.String())
	fmt.Printf("Keys match: %v\n\n", derivedPublicKey.String() == kp.PublicKey.String())

	// Demonstrate hex conversion
	privateHex := kp.PrivateKey.String()
	recoveredPrivateKey, err := keypair.PrivateKeyFromHex(privateHex)
	if err != nil {
		log.Fatalf("Failed to recover private key from hex: %v", err)
	}

	publicHex := kp.PublicKey.String()
	recoveredPublicKey, err := keypair.PublicKeyFromHex(publicHex)
	if err != nil {
		log.Fatalf("Failed to recover public key from hex: %v", err)
	}

	fmt.Println("Hex Conversion Test:")
	fmt.Printf("Private Key Recovered: %v\n", privateHex == recoveredPrivateKey.String())
	fmt.Printf("Public Key Recovered:  %v\n", publicHex == recoveredPublicKey.String())

	// Create a new key pair from existing keys
	newKeyPair := keypair.NewKeyPair(recoveredPrivateKey, recoveredPublicKey)
	fmt.Println("\nRecreated Key Pair:")
	fmt.Printf("Private Key: %s\n", newKeyPair.PrivateKey.String())
	fmt.Printf("Public Key:  %s\n", newKeyPair.PublicKey.String())
}